first pass

Hellowlol opened the pull request.

yozik04 commented on custom_components/nordpool/__init__.py:
```diff
@@ -64,23 +92,54 @@ async def _update(self, type_="today", dt=None):
+                else:
+                    _LOGGER.debug("Retrying request for %s", type_)
+                    p = partial(self._update, type_, dt)
+                    async_call_later(hass, 60 * 5, p)
```
I'd better create a loop function that would mostly sit in asyncio.sleep and periodically would call an update function. Depending on the update result that would configure next sleep cycle. But that is a code style preference.
Then update function would be much simpler and more readable.

Hellowlol replied:
Thanks for the suggestion! I'll use backoff for this. It will be something like this. https://replit.com/@Hellowlol/MoccasinSeparateDesignmethod#main.py

yozik04 replied:
New dependency? Do not see that homeassistant uses it.
Found this:
https://github.com/home-assistant/core/blob/9f8c0685e35da0ee4ba693395be5d2f77917ea3f/homeassistant/components/apple_tv/__init__.py#L215

Hellowlol replied:
Yes it is. It dont matter if core using it or not it will be installed if somebody uses this integration

yozik04 commented:
With the version from yesterday there is a bug that it keeps the same prices from yesterday.

janneojala commented on custom_components/nordpool/__init__.py:
```diff
@@ -4,18 +4,20 @@
 
 from .aio_price import AioPrices
 from .events import async_track_time_change_in_tz
+from .misc import test_valid_nordpooldata, test_valid_nordpolldata2, stock
```
test_valid_nordpolldata2 is not defined in misc.py

yozik04 commented:
2022-04-02 10:37:35 WARNING (MainThread) [homeassistant.components.sensor] Setup of sensor platform nordpool is taking over 10 seconds.
....
2022-04-02 10:41:26 WARNING (MainThread) [homeassistant.bootstrap] Waiting on integrations to complete setup: sensor.nordpool

Starts already for 4 minutes. something is wrong there...

Hellowlol commented:
Please enable the debug log and the log. The sensor can take a long time depending on the issues it has

yozik04 commented:
The integration made 69 requests to get data. i think it fails if prices are requested before midday (when new prices are published)
See log attached.
[nordpool-fix-1_full.log](https://github.com/custom-components/nordpool/files/8402532/nordpool-fix-1_full.log)

Hellowlol commented:
@yozik04 Thanks for the debug log! It was really helpful, the problem was that the retry method considers all "falsy" as a request for retrying. Ill fix that :)

yozik04 commented:
The method you used to make nordpool library async is very scary. I'd better throw a PR to add async to that library. =)

yozik04 commented:
BTW. Take a look to the latest version of that library. It already has async support.

yozik04 commented:
Ahh, you have just made a copy of that async version and modified... Ok, your code, your rules...

Hellowlol commented:
@yozik04 The API wasn't async when I made this, I was the one that sent an upstream PR to make it async. I kept my code to make sure that it worked. See PR https://github.com/kipe/nordpool/pull/8 and https://github.com/kipe/nordpool/pull/11 If you have a better way to make this async while supporting multiple async HTTP libraries, feel free so send a PR

Hellowlol commented:
@yozik04 I think it's ready to be merged. Can you test and leave it running until the prices for Monday should be available?

yozik04 commented:
@Hellowlol I will. Updating

yozik04 commented:
It still does 6 requests to nordpool on HA restart.

yozik04 commented:
Updating todays prices and Updating tomorrows prices.

Hellowlol commented:
@yozik04 thats intended. The api stores/get the data in cet. So if you requested data area in another time zone you wouldn’t get the correct prices. That’s why we request data for 3 days and extract what we need.

yozik04 commented:
It is ok to fetch 3. But why it fetches 3 requests two times.

Hellowlol commented:
Its one request for today and one for tomorrow.

yozik04 commented:
`Found infinty invalid data in area SYS`
Why it checks SYS if I am interested in EE only?

Looks like in my region prices appear at 15:00 UTC+3. For some reason it started requesting at 14:00, which is 11 UTC.

But overall it works.

yozik04 commented:
Ahh, my father has restarted HA in the morning. He told it showed today and tomorrow same prices...

Hellowlol commented:
@yozik04 can you post the log? I want to check why it started requesting data then

yozik04 commented:
Sure. Here you go.

yesterday evening + todays morning till 8 am
[nordpool-2022-04-02_03.log](https://github.com/custom-components/nordpool/files/8404593/nordpool-2022-04-02_03.log)

todays 8 am+
[nordpool-2022-04-03.log](https://github.com/custom-components/nordpool/files/8404594/nordpool-2022-04-03.log)

yozik04 commented on custom_components/nordpool/misc.py:
```diff
@@ -26,29 +28,30 @@
-        _LOGGER.debug("No data is available yet.")
+        _LOGGER.debug("No data is available yet. Stopping retries")
         return False
     return not_(value)
```
Doesn't `return not value` work?

Hellowlol replied:
Yeah, that will work too. 👍

yozik04 commented on custom_components/nordpool/misc.py:
```diff
@@ -26,29 +28,30 @@
 
 
-def predicate(value):
+def predicate(value: Union[bool, None]) -> bool:
```
Type hint. I see only bool output here.

Hellowlol replied:
The output can only be bool, but the input might be None, True, False

Hellowlol commented:
@yozik04 I couldn't find any obvious reason for why tomorrows data wasn't cleared. I have added some better logging and fixed a potential race condition. Please update and let me know if you have any issues.

yozik04 commented on custom_components/nordpool/__init__.py:
```diff
@@ -91,9 +94,8 @@ async def _update(self, *args, type_="today", dt=None):
         @backoff.on_predicate(backoff.expo, predicate=predicate, logger=_LOGGER)
         @backoff.on_exception(backoff.expo, aiohttp.ClientError, logger=_LOGGER)
-        async def really_update(currency, end_date):
+        async def really_update(currency: str, end_date: datetime) -> Union[bool, None]:
```
Doesn't it return only bool?

Hellowlol replied:
It can also be None

yozik04 reviewed (commented):
Will pull and try this version

yozik04 commented:
I will drop an idea here how I would do the whole thing.

I'd create a class that would hold all fetched valid values (yesterday, today, tomorrow) in a single dictionary (datetime as a key). Let's name it PriceData That class would have methods:

- `get_todays_data` - would scan the dictionary and output todays's prices
- `get_tomorrows_data` - would scan the dictionary and output tomorrow's prices
- `has_todays_data` - would tell if dictionary has todays data
- `has_tomorrows_data` - would tell if dictionary has tomorrows data
- `clear_expired` - would scan the dictionary and drop data that is already expired
- `store_data` - injects the data to the dictionary (actually a nice place to validate the data as well, and inject only valid values)
- `get_current_price` - probably it is also needed, you know better.

Code out of PriceData class would query nordpool for today's data only if `PriceData.has_todays_data` returns `False`.
Same with tomorrow but should also verify that time is adequate for the request. If fetch was successful and data is valid we can still reuse `has_todays_data`, `has_tomorrows_data`

Every day something needs to call `PriceData.clear_expired` as well (probably run it before every request to nordpool).

That would be my design. I would even cover PriceData with some unit tests.
But I do not really want to go and change your code. =) No offence.

Hellowlol commented:
> I will drop an idea here how I would do the whole thing.
> `
> I'd create a class that would hold all fetched valid values (yesterday, today, tomorrow) in a single dictionary (datetime as a key). Let's name it PriceData That class would have methods:

Thanks for your suggestion, that is pretty much how this is designed today with some small changes to be able to handle currencies and we also need to clear the sensors data in addition to `NordpoolData` as we can't do IO in a property.
> 
> * `get_todays_data` - would scan the dictionary and output todays's prices `today`
> * `get_tomorrows_data` - would scan the dictionary and output tomorrow's prices ´tomorrow`
> * `has_todays_data` - would tell if dictionary has todays data | Does not exist
> * `has_tomorrows_data` - would tell if dictionary has tomorrows data | Does not exist
> * `clear_expired` - would scan the dictionary and drop data that is already expired `new_day_cb`
> * `store_data` - injects the data to the dictionary (actually a nice place to validate the data as well, and inject only valid values) `_update`
> * `get_current_price` - probably it is also needed, you know better.
> 
> Code out of PriceData class would query nordpool for today's data only if `PriceData.has_todays_data` returns `False`. Same with tomorrow but should also verify that time is adequate for the request. If fetch was successful and data is valid we can still reuse `has_todays_data`, `has_tomorrows_data`
> 
> Every day something needs to call `PriceData.clear_expired` as well (probably run it before every request to nordpool).
> 
> That would be my design. I would even cover PriceData with some unit tests. But I do not really want to go and change your code. =) No offence.

I'll accept every PR that makes the sensor more reliable as long as none of the current features are missed. 😊

Hellowlol commented:
Ill add some tests

yozik04 commented:
At night both prices day prices became unavailable with new version. Will send logs later.

yozik04 commented:
Todays log: [nordpool-2022-04-04.log](https://github.com/custom-components/nordpool/files/8407138/nordpool-2022-04-04.log)

yozik04 commented:
yesterday's logs were lost unfortunately.

Hellowlol commented:
Thanks, that was helpful. 👍 The server was restarted after midnight so there isnæt any available data for tomorrow, it will try one then, stop. However, it also didn a new http call to get todays data. There was invalid values in FI result, so det requests for retried over and over until the setup of the senor failed. We validate on currency as the NordpoolData don't know what data the user wants

yozik04 commented:
[nordpool_2022-04-04_05.log](https://github.com/custom-components/nordpool/files/8417625/nordpool_2022-04-04_05.log)
Worked fine today on version before tests. Posting log for reference.

yozik04 commented:
I can confirm the last one is working as expected.

Hellowlol commented:
@yozik04 Can you do one last test? I was thinking of making a new release this weekend.

yozik04 commented:
Pulled latest changes.
I have 00:37 on the clock now and it failed to download/update the data. Found infinity inside FI... Will see if it will recover during the night.
As before, at the night time it fails. Due to Nordpool returning junk or an other reason. Anyway I will send logs tomorrow. I have not digged what is in the returned XML yet. There are some valid values for sure.

Hellowlol commented:
The latest update should only verify the area that the user requested. If the user has multiple sensors using the same currency, then all of the data for the selected areas have to be valid.

yozik04 commented:
Logs from version 34be90f 
[nordpool-2022-04-07_night.log](https://github.com/custom-components/nordpool/files/8443798/nordpool-2022-04-07_night.log)

JSON files that were fetched at night:
[08-04-2022.txt](https://github.com/custom-components/nordpool/files/8443822/08-04-2022.txt)
[07-04-2022.txt](https://github.com/custom-components/nordpool/files/8443823/07-04-2022.txt)
[06-04-2022.txt](https://github.com/custom-components/nordpool/files/8443824/06-04-2022.txt)

```
cat 06-04-2022.txt | jq -c '[.data.Rows[].Columns | map(select(.Name == "EE"))[].Value]'
["81,73","72,17","69,23","71,48","74,79","74,13","87,07","136,61","151,41","95,96","90,42","87,16","86,68","84,67","84,90","86,91","87,68","87,63","87,91","104,52","111,75","54,68","45,08","27,35","27,35","151,41","85,08","94,65","83,40","59,72"]

cat 07-04-2022.txt | jq -c '[.data.Rows[].Columns | map(select(.Name == "EE"))[].Value]'
["81,73","72,17","69,23","71,48","74,79","74,13","87,07","136,61","151,41","95,96","90,42","87,16","86,68","84,67","84,90","86,91","87,68","87,63","87,91","104,52","111,75","54,68","45,08","27,35","27,35","151,41","85,08","94,65","83,40","59,72"]

cat 08-04-2022.txt | jq -c '[.data.Rows[].Columns | map(select(.Name == "EE"))[].Value]'
["-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-","-"]
```

As you can see 6th April's data is absolutely equal to 7th April, what is bullshit. 8th is not available.

Hellowlol commented:
Thanks again! Clearly it’s bug in the retry. However I have no idea why the values for today and tomorrow was the same. How if you get the json? Did you do manual request and download it?   The raw json isn’t logged in this integration.

yozik04 commented:
I have copied urls from the HA log and used wget at that time.

poffe commented:
Hi! Just a friendly poke to see how this is progressing? I've seen this issue a couple of times so that's the reason. Great work btw.

Hellowlol commented:
I haven't forgotten this. I just got frustrated with the API returning junk :P It will get merged sometime in the future :D
